src/txmempool.h

```diff
@@ -53,13 +57,66 @@ class CTxMemPoolEntry
 
     const CTransaction& GetTx() const { return this->tx; }
     double GetPriority(unsigned int currentHeight) const;
+    double GetCurrentPriority() const { return dCurrentPriority; }
     CAmount GetFee() const { return nFee; }
+    CFeeRate GetFeeRate() const { return feeRate; }
     size_t GetTxSize() const { return nTxSize; }
     int64_t GetTime() const { return nTime; }
     unsigned int GetHeight() const { return nHeight; }
     bool WasClearAtEntry() const { return hadNoDependencies; }
+    void recalcCurrentPriority (unsigned int height) { dCurrentPriority = GetPriority(height); }
 };
 
+// extracts a TxMemPoolEntry's transaction hash
+struct mempoolentry_txid
+{
+    typedef uint256 result_type;
+    result_type operator() (const CTxMemPoolEntry &entry) const
+    {
+        return entry.GetTx().GetHash();
+    }
+};
+
+class CompareTxMemPoolEntryByPriority
+{
+public:
+    bool operator()(const CTxMemPoolEntry& a, const CTxMemPoolEntry& b)
+    {
+        if (a.GetCurrentPriority() == b.GetCurrentPriority())
+            return a.GetFeeRate() > b.GetFeeRate();
+        return a.GetCurrentPriority() > b.GetCurrentPriority();
+    }
+};
+
+class CompareTxMemPoolEntryByFee
+{
+public:
+    bool operator()(const CTxMemPoolEntry& a, const CTxMemPoolEntry& b)
+    {
+        if (a.GetFeeRate() == b.GetFeeRate())
+            return a.GetCurrentPriority() > b.GetCurrentPriority();
+        return a.GetFeeRate() > b.GetFeeRate();
+    }
+};
+
+typedef boost::multi_index_container<
+    CTxMemPoolEntry,
+    boost::multi_index::indexed_by<
+        // sorted by txid
+        boost::multi_index::ordered_unique<mempoolentry_txid>,
+        // sorted by fee rate
+        boost::multi_index::ordered_non_unique<
+            boost::multi_index::identity<CTxMemPoolEntry>,
+            CompareTxMemPoolEntryByFee
+        >,
+        // sorted by priority
+        boost::multi_index::ordered_non_unique<
+            boost::multi_index::identity<CTxMemPoolEntry>,
+            CompareTxMemPoolEntryByPriority
+        >
+    >
+> ordered_transaction_set;
```

Comment:
It is nontrivial to make boost::multi_index_container work with salted hash
functions (you want to avoid the ability for network players to produce
transactions that collide in the hashtable). I've tried myself, so I think
it's reasonable to first get this in without. One of the advantages of
boost::multi_index_container is that changing the indexes should be easy.